src/daft-dsl/src/functions/scalar.rs

```diff
@@ -84,3 +85,337 @@ impl Display for ScalarFunction {
+                        _ => Err(common_error::DaftError::ValueError(format!("function '{}' expects 2 argument, instead received {}", $func_name , inputs.len())))
+                    }
+                }
+            }
+
+            pub fn [<$func_name:lower>](input: $crate::ExprRef, arg0: $crate::ExprRef, arg1: $crate::ExprRef) -> $crate::ExprRef {
+                $crate::functions::ScalarFunction::new([<$func_name:camel Function>] {}, vec![input,arg0, arg1]).into()
+            }
+
+            #[cfg(feature = "python")]
+            #[pyo3::pyfunction]
+            #[pyo3(name = $func_name)]
+            pub fn [<py_ $func_name:lower>](expr: daft_dsl::python::PyExpr, arg0: daft_dsl::python::PyExpr, arg1: daft_dsl::python::PyExpr) -> pyo3::prelude::PyResult<daft_dsl::python::PyExpr> {
+                Ok([<$func_name:lower>](expr.into(), arg0.into(), arg1.into()).into())
+            }
+        }
+    };
+
+}
+
+/// generates a scalar UDF implementation
+/// Similar to `make_udf_function`, but with parameterized inputs
+/// # Example
+/// ```rust,ignore
+///
+/// make_parameterized_udf_function!{
+///     name: "my_udf",
+///     params: (query: String, param2: u64),
+///     to_field: (inputs, schema) {
+///       // implementation
+///     },
+///     evaluate: (self, inputs) {
+///       // self.query
+///       // self.param2
+///       // implementation
+///     }
+/// }
+/// ```
+/// generates the following code
+/// ```rust,ignore
+/// #[derive(Debug, Clone, serde::Serialize, serde::Deserialize, PartialEq, Eq, Hash)]
+/// pub struct MyUDF {
+///    query: String,
+///    param2: u64
+/// }
+///
+/// impl ScalarUDF for MyUDF {
+///    fn as_any(&self) -> &dyn std::any::Any { self }
+///    fn name(&self) -> &'static str { "my_udf" }
+///
+///    fn to_field(&self, inputs: &[ExprRef], schema: &daft_core::schema::Schema) -> common_error::DaftResult<daft_core::datatypes::Field> {
+///      match inputs {
+///        [input0] => { // implementation }
+///        _ => Err(common_error::DaftError::InvalidInput("my_udf".to_string()))
+///      }
+///    }
+///
+///    fn evaluate(&self, inputs: &[daft_core::series::Series]) -> common_error::DaftResult<daft_core::series::Series> {
+///      match inputs {
+///        [input0] => { // implementation }
+///        _ => Err(common_error::DaftError::InvalidInput("my_udf".to_string()))
+///      }
+///    }
+/// }
+/// fn my_udf(input: ExprRef, query: String, param2: u64) -> ExprRef {
+///    ScalarFunction::new(MyUDF {query, param2}, vec![input, arg]).into()
+/// }
+///
+/// #[cfg(feature = "python")]
+/// #[pyo3::pyfunction]
+/// #[pyo3(name = "my_udf")]
+/// pub fn py_my_udf(expr: PyExpr, query: String, param2: u64) -> PyResult<PyExpr> {
+///   my_udf(expr.into(), query, param2).into()
+/// }
+/// ```
+#[macro_export]
+macro_rules! make_parameterized_udf_function {
+    (
+        name: $func_name:expr,
+        params: ($($param:ident: $param_type:ident)*),
```

Comment:
Small nit:

Maybe we have `params: ($($param:ident: $param_type:ident),*$(,)?)`?

That change would just allow for the parameters to be comma-delimited (with an optional comma at the end).